src/logic/map_objects/tribes/militarysite.cc

```diff
@@ -727,31 +727,47 @@ void MilitarySite::act(Game& game, uint32_t const data) {
 		update_soldier_request();
 	}
 
+	// Heal soldiers
 	if (nexthealtime_ <= timeofgame) {
-		uint32_t total_heal = descr().get_heal_per_second();
-		std::vector<Soldier*> soldiers = soldier_control_.present_soldiers();
+		const uint32_t total_heal = descr().get_heal_per_second();
 		uint32_t max_total_level = 0;
 		float max_health = 0;
-		Soldier* soldier_to_heal = 0;
-
-		for (uint32_t i = 0; i < soldiers.size(); ++i) {
-			Soldier* s = soldiers[i];
-
-			// The healing algorithm is:
-			// * heal soldier with highest total level
-			// * heal healthiest if multiple of same total level exist
-			if (s->get_current_health() < s->get_max_health()) {
-				if (0 == soldier_to_heal || s->get_total_level() > max_total_level ||
-				    (s->get_total_level() == max_total_level &&
-				     s->get_current_health() / s->get_max_health() > max_health)) {
-					max_total_level = s->get_total_level();
-					max_health = s->get_current_health() / s->get_max_health();
-					soldier_to_heal = s;
+		Soldier* soldier_to_heal = nullptr;
+
+		for (Soldier* soldier : soldier_control_.stationed_soldiers()) {
+			if (soldier->get_current_health() < soldier->get_max_health()) {
+				if (is_present(*soldier)) {
+					// The healing algorithm for present soldiers is:
+					// * heal soldier with highest total level
+					// * heal healthiest if multiple of same total level exist
+					if (soldier_to_heal == nullptr || soldier->get_total_level() > max_total_level ||
+					    (soldier->get_total_level() == max_total_level &&
+					     soldier->get_current_health() / soldier->get_max_health() > max_health)) {
+						max_total_level = soldier->get_total_level();
+						max_health = soldier->get_current_health() / soldier->get_max_health();
+						soldier_to_heal = soldier;
+					}
+				} else if ((soldier->get_battle() == nullptr ||
+				            soldier->get_battle()->opponent(*soldier) == nullptr) &&
+				           !get_economy(WareWorker::wwWORKER)->warehouses().empty()) {
+					// Somewhat heal soldiers in the field that are not currently engaged in fighting an
+					// opponent,
+					// but only if there is a warehouse connected.
```

Comment:
Could you make these two line one, please? My OCD is killing me :-)

Reply:
Done.